NullPointerException on getting excluded/included groups

TestNg Version 6.9.13.6

My testng.xml file contains block `<groups></groups>` with some defined excluded groups in. I'm running the tests by using Maven. Just on starting I've got the following NPE:

```
Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project ui-test: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: There was an error in the forked process] 
org.testng.TestNGException: java.lang.NullPointerException
at org.testng.TestNG.initializeSuitesAndJarFile(TestNG.java:325)
at org.testng.TestNG.run(TestNG.java:1081)
at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:281)
at org.apache.maven.surefire.testng.TestNGXmlTestSuite.execute(TestNGXmlTestSuite.java:75)
at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:121)
at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290) 
at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: java.lang.NullPointerException
at org.testng.xml.XmlSuite.getExcludedGroups(XmlSuite.java:1072)
at org.testng.xml.XmlTest.getExcludedGroups(XmlTest.java:227)
at org.testng.internal.OverrideProcessor.process(OverrideProcessor.java:37)
at org.testng.xml.Parser.parse(Parser.java:225)
at org.testng.TestNG.initializeSuitesAndJarFile(TestNG.java:305)
```

Everything works OK on the previous older version (e.g. 6.9.8)

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,5 +1,5 @@
 ﻿Current
-
+Fixed: GITHUB-1180: NullPointerException on getting excluded/included groups (Krishnan Mahadevan)
 Fixed: GITHUB-1064: Incorrect logging of parallel mode of a test
 Fixed: GITHUB-1178: Halt execution when invalid testname is provided. (Krishnan Mahadevan)
 Fixed: GITHUB-1139: DataProvider could support Object[] as a valid return type (Julien Herr)

--- a/src/main/java/org/testng/xml/XmlGroups.java
+++ b/src/main/java/org/testng/xml/XmlGroups.java
@@ -1,10 +1,10 @@
 package org.testng.xml;
 
-import java.util.List;
-
 import org.testng.collections.Lists;
 import org.testng.reporters.XMLStringBuffer;
 import org.testng.xml.dom.Tag;
+
+import java.util.List;
 
 import static org.testng.collections.CollectionUtils.hasElements;
 
@@ -63,7 +63,11 @@ public class XmlGroups {
       xsb.getStringBuffer().append(d.toXml(indent2));
     }
 
-    xsb.getStringBuffer().append(m_run.toXml(indent2));
+      if (null != m_run) {
+          //XmlRun is optional and is not always available. So check if its available before running toXml()
+          xsb.getStringBuffer().append(m_run.toXml(indent2));
+      }
+
 
     for (XmlDependencies d : m_dependencies) {
       xsb.getStringBuffer().append(d.toXml(indent2));

--- a/src/main/java/org/testng/xml/XmlSuite.java
+++ b/src/main/java/org/testng/xml/XmlSuite.java
@@ -1035,7 +1035,7 @@ public class XmlSuite implements Serializable, Cloneable {
   public List<String> getIncludedGroups() {
     if (m_parentSuite != null) {
       return m_parentSuite.getIncludedGroups();
-    } else if (m_xmlGroups != null) {
+    } else if (m_xmlGroups != null && (m_xmlGroups.getRun() != null)) {
       return m_xmlGroups.getRun().getIncludes();
     } else {
       // deprecated
@@ -1068,7 +1068,7 @@ public class XmlSuite implements Serializable, Cloneable {
   public List<String> getExcludedGroups() {
     if (m_parentSuite != null) {
       return m_parentSuite.getExcludedGroups();
-    } else if (m_xmlGroups != null) {
+    } else if (m_xmlGroups != null && (m_xmlGroups.getRun() != null)) {
       return m_xmlGroups.getRun().getExcludes();
     } else {
       return m_excludedGroups;

--- a/src/test/java/org/testng/xml/XmlSuiteTest.java
+++ b/src/test/java/org/testng/xml/XmlSuiteTest.java
@@ -1,0 +1,33 @@
+package org.testng.xml;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import test.SimpleBaseTest;
+
+import java.util.Arrays;
+
+public class XmlSuiteTest extends SimpleBaseTest {
+
+    @Test
+    public void testIncludedAndExcludedGroups() {
+        XmlSuite suite = new XmlSuite();
+        suite.addIncludedGroup("foo");
+        suite.addExcludedGroup("bar");
+        Assert.assertEquals(Arrays.asList("foo"), suite.getIncludedGroups());
+        Assert.assertEquals(Arrays.asList("bar"), suite.getExcludedGroups());
+    }
+
+    @Test
+    public void testIncludedAndExcludedGroupsWithRun() {
+        XmlRun xmlRun = new XmlRun();
+        xmlRun.onInclude("foo");
+        xmlRun.onExclude("bar");
+        XmlGroups groups = new XmlGroups();
+        groups.setRun(xmlRun);
+        XmlSuite suite = new XmlSuite();
+        suite.setGroups(groups);
+        Assert.assertEquals(Arrays.asList("foo"), suite.getIncludedGroups());
+        Assert.assertEquals(Arrays.asList("bar"), suite.getExcludedGroups());
+    }
+
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -691,6 +691,7 @@
   <test name="XML">
     <classes>
       <class name="org.testng.xml.SuiteXmlParserTest" />
+      <class name="org.testng.xml.XmlSuiteTest"/>
     </classes>
   </test>
 
@@ -775,5 +776,6 @@
       <class name="test.thread.parallelization.ParallelByMethodsMultipleSuitesTestsClassesTest"/>
     </classes>
   </test>
+
 </suite>
 

